### DISCUSSION TITLE
Defining creep rate with a custom function

### DISCUSSION THREAD

**Discussion by AndrewFalkowski** [idaholab/moose#22879] [Q&A Modules: Solid mechanics] [answered]
I have creep rate data as a function of temperature and shear stress in a table. After importing this data as a bilinear interpolation function is there a way to utilize this function for modeling creep? Or would a custom kernel need to be written for this?

**Comment by GiudGiud** [comment, marked as answer]
@bwspenc @recuero

**Reply by GiudGiud** [reply]
So something similar to this
https://mooseframework.inl.gov/source/materials/PowerLawCreepStressUpdate.html
but not power law?

**Reply by AndrewFalkowski** [reply]
Yes, in this case I don't have access to fitted power law parameters and instead just have creep rates that result from a combination of shear stress and temperature. The data covers several different creep regimes, which would make simple power law fitting inapplicable.

**Reply by recuero** [reply]
If your use case doesn't use a power law and directly prescribes the creep rate as a function of some variables, I'd create a new creep object which takes a function with a number of parameters, some of them may be optional. So, I don't think you need a kernel but a material property.

**Reply by AndrewFalkowski** [reply]
Would you be willing to provide a simple example of the implementation of that in a MOOSE input file?

**Reply by recuero** [reply]
I don't think we currently have that capability in MOOSE (I may be missing some options though). So I don't think you can prescribe that from the input file; you'd need an a creep object that takes `Function`s as input parameters and operate on them with coupled variables to return the creep rate that your data represents. You could define a function in the input file that takes parameters as temperature and then pass the function to such an object. There are objects that take functions, such as: `ComputeDynamicFrictionalForceLMMechanicalContact` with 

```
params.addParam<FunctionName>(
      "function_friction",
      "Coupled function to evaluate friction with values from contact pressure and relative "
      "tangential velocities (from the previous step).");

**Reply by AndrewFalkowski** [reply]
Pulling a variable dependent function into an object seems simple enough. 

I assume that any creep object is going to follow a similar structure to the other `StressUpdate` objects in order to work with `ComputeMultipleInelasticStress`.  Can I follow the structure of the `ComputeMultipleInelasticStress` source file as a reference or would some of the quantities computed in that file be unnecessary in my implementation?

**Reply by recuero** [reply]
I would follow the same structure as other stress update objects that work with ComputeMultipleInelasticStress. There are a number of capabilities associated with that object: substepping, management of various inelasticity sources, tangent operator, etc.

**Reply by AndrewFalkowski** [reply]
Gotcha thank you! I was able to build the custom object and get reasonable values for my creep rate, but I am struggling to get the simulation to converge with the object active. Is there an effective way or recommended approach to debug this?

```
Exceeded maximum iterations in material return mapping iterations.
At element 0 _qp=0 Coordinates (x,y,z)=(-0.000781231, 0.0013792, 0.00186622) block=1
In 1000 iterations the residual went from 1.53452e+07 to 3.35603e+11 in 'FrostPowerLawCreep'.
```

**Reply by AndrewFalkowski** [reply]
To follow up on this, I have pivoted from pulling from a lookup table and am instead using fitted function to compute my creep rate in the form: $$A * \sigma^n * exp\left(\frac{-Q_1}{RT}\right) + B * \sigma^m * exp\left(\frac{-Q_2}{RT}\right)$$

I believe I have implemented it correctly, but I continue to get the above error when running it. Although the residual just is often more extreme as shown below.
```
Exceeded maximum iterations in material return mapping iterations.
At element 0 _qp=1 Coordinates (x,y,z)=(-0.0010648, 0.00128243, 0.00186622) block=1
In 1000 iterations the residual went from 2.02083e+48 to -2.39293e-10 in 'FrostPowerLaw_creep'.
```
When using `use_substep=true` I am returned a the message `The number of substeps computed exceeds the maximum_number_substeps. The system time step will be cut.` , which is then followed by the following error message:
```
  Elem Information
   id()=0, unique_id()=5704, processor_id()=0
   type()=HEX8
   dim()=3
   n_nodes()=8
   mapping=LAGRANGE_MAP
    0  Node id()=0, processor_id()=0, Point=(x,y,z)=(0.00739408, -0.00908827, 0.0165142)
    DoFs=(0/0/0) (0/1/1) (0/2/2) (1/0/0) 
    1  Node id()=1, processor_id()=0, Point=(x,y,z)=(0.00804161, -0.0089018, 0.0167091)
    DoFs=(0/0/3) (0/1/4) (0/2/5) (1/0/1) 
    2  Node id()=2, processor_id()=0, Point=(x,y,z)=(0.00787107, -0.00843567, 0.0162784)
    DoFs=(0/0/6) (0/1/7) (0/2/8) (1/0/2) 
    3  Node id()=3, processor_id()=0, Point=(x,y,z)=(0.00731591, -0.00865312, 0.0160966)
    DoFs=(0/0/9) (0/1/10) (0/2/11) (1/0/3) 
    4  Node id()=4, processor_id()=0, Point=(x,y,z)=(0.00989949, -0.0112232, 0.0196849)
    DoFs=(0/0/12) (0/1/13) (0/2/14) (1/0/4) 
    5  Node id()=5, processor_id()=0, Point=(x,y,z)=(0.0104961, -0.0110342, 0.0199122)
    DoFs=(0/0/15) (0/1/16) (0/2/17) (1/0/5) 
    6  Node id()=6, processor_id()=0, Point=(x,y,z)=(0.0103128, -0.0105176, 0.0194098)
    DoFs=(0/0/18) (0/1/19) (0/2/20) (1/0/6) 
    7  Node id()=7, processor_id()=0, Point=(x,y,z)=(0.00980942, -0.010741, 0.0191968)
    DoFs=(0/0/21) (0/1/22) (0/2/23) (1/0/7) 
   n_sides()=6
    neighbor(0)=669
    neighbor(1)=1
    neighbor(2)=41
    neighbor(3)=4
    neighbor(4)=30
    neighbor(5)=nullptr
   hmin()=0.000590468, hmax()=0.00550838
   volume()=-2.85096e-11
   active()=1, ancestor()=0, subactive()=0, has_children()=0
   parent()=nullptr
   level()=0, p_level()=0
   refinement_flag()=DO_NOTHING
   p_refinement_flag()=DO_NOTHING
   DoFs=(1/1/863) (1/1/864) (1/1/865) (1/1/866) (1/1/867) (1/1/868) (1/1/869) (1/1/870) (1/1/871) (1/1/872) (1/2/873) (1/2/874) (1/2/875) (1/2/876) (1/2/877) (1/2/878) (1/2/879) (1/2/880) (1/2/881) (1/2/882) (1/3/883) (1/3/884) (1/3/885) (1/3/886) (1/3/887) (1/3/888) (1/3/889) (1/3/890) (1/3/891) (1/3/892) (1/4/893) (1/4/894) (1/4/895) (1/4/896) (1/4/897) (1/4/898) (1/4/899) (1/4/900) (1/4/901) (1/4/902) (1/5/903) (1/5/904) (1/5/905) (1/5/906) (1/5/907) (1/5/908) (1/5/909) (1/5/910) (1/5/911) (1/5/912) (1/6/913) (1/6/914) (1/6/915) (1/6/916) (1/6/917) (1/6/918) (1/6/919) (1/6/920) (1/6/921) (1/6/922) 
ERROR: negative Jacobian -2.0273e-12 at point (x,y,z)=(0.00773864, -0.00925779, 0.0168458) in element 0
[0] ../src/fe/fe_map.C, line 1173, compiled Aug 25 2022 at 10:22:39
We caught a libMesh error in ThreadedElementLoopBase
Cannot take square root of a number less than or equal to zero in the calculation of C1 for the Rashid approximation for the rotation tensor. This zero or negative number may occur when elements become heavily distorted.
```

Any advice on how to debug or remedy this would be greatly appreciated.

**Reply by recuero** [reply]
>  am instead using fitted function to compute my creep rate in the form

You are probably then using MOOSE's capabilities, which allow to combine power law functions.

>In 1000 iterations the residual went from 2.02083e+48 to -2.39293e-10 in 'ZrC_creep'

It seems that your residual went down 58 orders of magnitude. I'd probably modify the relative or absolute tolerance to allow for that reduction in residual to be acceptable.

> The number of substeps computed exceeds the maximum_number_substeps. The system time step will be cut.

The number of substeps will depend on the criterion used to perform substepping. If the tolerance you used for that is okay, this message probably means your time step is too large.

**Reply by AndrewFalkowski** [reply]
> You are probably then using MOOSE's capabilities, which allow to combine power law functions.

Would using multiple instances of `PowerLawCreepStressUpdate` with `ADComputeMultipleInelasticStress` result in the same additive creep rate? I'm not sure I fully understand how `ADComputeMultipleInelasticStress` determines the contributions of different models.

> I'd probably modify the relative or absolute tolerance to allow for that reduction in residual to be acceptable.

That seems like a pretty extraordinary drop in the residual, or is that reasonable in high strain plasticity models?

**Reply by recuero** [reply]
> Would using multiple instances of PowerLawCreepStressUpdate with ADComputeMultipleInelasticStress result in the same additive creep rate?

Yes, it will. `ADComputeMultipleInelasticStress` would also allow adding other inelastic models such as plastic ones, if needed.

> That seems like a pretty extraordinary drop in the residual, or is that reasonable in high strain plasticity models?

Without knowing much, I'd be satisfied with that reduction in the residual. You can debug it and see if it stagnates at e-10, which is probably the case. It is possible that the system is reaching some sort of machine precision limit. I don't know what units you are using, so I don't have a good answer as to what value for the initial residual would be reasonable. You can always play around a bit with the iteration debug output (`internal_solve_full_iteration_history = true`), which is supported in those classes.

**Reply by AndrewFalkowski** [reply]
Great, that seems to have resolved my issues. One last question, when defining multiple creep equations, I am getting an error saying `The following material properties are declared on block 0 by multiple materials...`. When applying the recommended solution in [#22714](https://github.com/idaholab/moose/discussions/22714) I am then getting an error that the necessary properties aren't being defined:
```
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'creep_strain', requested by 'creep_strain_xx_all' is not defined on block 0
Material property 'creep_strain', requested by 'creep_strain_xx_all_face' is not defined on block 0
Material property 'creep_strain', requested by 'creep_strain_xx_all_neighbor' is not defined on block 0
Material property 'creep_strain', requested by 'creep_strain_yy_all' is not defined on block 0
Material property 'creep_strain', requested by 'creep_strain_yy_all_face' is not defined on block 0
Material property 'creep_strain', requested by 'creep_strain_yy_all_neighbor' is not defined on block 0
Material property 'creep_strain', requested by 'creep_strain_zz_all' is not defined on block 0
Material property 'creep_strain', requested by 'creep_strain_zz_all_face' is not defined on block 0
Material property 'creep_strain', requested by 'creep_strain_zz_all_neighbor' is not defined on block 0
```
Relevant code shown below:
```
  [material_elasticity]
    type = ADComputeVariableIsotropicElasticityTensor
    youngs_modulus = E_material
    poissons_ratio = v_material
    block = 0
  []
  [material_plasticity]
    type = ADComputeMultipleInelasticStress
    block = 0
    inelastic_models ='ht_power_creep lt_power_creep'
  []
  [ht_power_creep]
    type = ADPowerLawCreepStressUpdate
    coefficient = 0.002241784250895549
    n_exponent = 6.999542917225692
    activation_energy = 464.62863657297504
    temperature = temperature
    base_name = 'ht_'
  []
  [lt_power_creep]
    type = ADPowerLawCreepStressUpdate
    coefficient = 60028109.64391723
    n_exponent = 5.004953887213269
    activation_energy = 716.7967951872396
    temperature = temperature
    base_name = 'lt_'
  []
```

**Reply by recuero** [reply]
Since you are righly using the base name, the creep strain material property gets the base name pretended, i.e. 'lt_'. You may be trying to output creep strains by the wrong name. It won't be just 'creep_strain...', but 'lt_creep_strain...'.

**Reply by AndrewFalkowski** [reply]
Ah, that makes sense; however, those defined variable names don't seem to be available when I try and call them within `[Modules/TensorMechanics/Master]` or through a post-processor. 

```
[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    incremental = true
    add_variables = true
    use_automatic_differentiation = true
    automatic_eigenstrain_names = true
    material_output_order = SECOND
    generate_output = 'vonmises_stress max_principal_stress min_principal_stress ht_creep_strain_xx'
  []
[]
```

Running this produces the error: `Unable to add output Material for 'ht_creep_strain_xx'`

What am I missing here?

**Reply by recuero** [reply]
Could you add those particular creep strains using the tandem auxiliary variable/auxiliary kernel. The action may only know about generic material names.

**Reply by AndrewFalkowski** [reply]
I'm not having much luck declaring through the aux system, but my implementation may be incorrect. I am getting an error that the property of interest doesn't exist on the block. Is there a way to check the names of all properties produced by a given material object?

**Reply by GiudGiud** [reply]
in the [Debug] block you can turn on show_material_properties = true

**Reply by AndrewFalkowski** [reply]
Ah ok, is there an example of how to pull one of the listed variables into a variable that can be visualized in the output?

**Reply by GiudGiud** [reply]
so in the Material that defines that property, you set `outputs = exodus` (or the name of your exodus output)

**Reply by AndrewFalkowski** [reply]
Great, thank you!
